buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -907,39 +1051,58 @@ private boolean allocateWithoutLock(int size, int maxCapacity, AdaptiveByteBuf b
         }
 
         private boolean allocate(int size, int maxCapacity, AdaptiveByteBuf buf, boolean reallocate) {
-            int startingCapacity = chunkController.computeBufferCapacity(size, maxCapacity, reallocate);
             Chunk curr = current;
             if (curr != null) {
-                // We have a Chunk that has some space left.
-                int remainingCapacity = curr.remainingCapacity();
-                if (remainingCapacity > startingCapacity) {
-                    curr.readInitInto(buf, size, startingCapacity, maxCapacity);
-                    // We still have some bytes left that we can use for the next allocation, just early return.
-                    return true;
-                }
-
-                // At this point we know that this will be the last time current will be used, so directly set it to
-                // null and release it once we are done.
-                current = null;
-                if (remainingCapacity >= size) {
-                    try {
-                        curr.readInitInto(buf, size, remainingCapacity, maxCapacity);
+                if (curr instanceof SizeClassedChunk) {
```

Comment:
Same as https://github.com/netty/netty/pull/15571/files#r2347893958

The 2 logic seems so different that IDK if it's clear to have both grouped here